### PR TITLE
[Debugger] Adding debupgy as the ray debugger

### PR DESCRIPTION
**Pull request by brycehuang30**
## Why are these changes needed?

* enabled debugpy as the ray debugger for breakpoint and post_mortem debugging
* added flag RAY_DEBUG=1 to enable debugpy. If RAY_DEBUG is not set and RAY_PDB is set, then rpdb will be used.
* used state api to save worker debugging port.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
 I manually tested that the task can be paused and it awaits connection from debugger client. Tested tasks running on worker nodes and head node. I also tested post-mortem debugging is working.

### CODE REVIEW

**Comment by sofianhnaide** [comment]
The files do matches the diff based on my review, please verify end to end working before merging.

**Comment by brycehuang30** [comment]
I manually tested that the task can be paused and it awaits connection from debugger client. Tested tasks running on worker nodes and head node. I also tested post-mortem debugging is working

**Review by sofianhnaide — APPROVED** [review]

File: python/setup.py [hunk 1]
```diff
@@ -342,6 +342,7 @@ def get_packages(self):
         "aiosignal",
         "frozenlist",
         "requests",
+        "debugpy",
```
Comment (richardliaw) [inline python/setup.py:345]: We probably shouldn't install this by default; can we instead do a soft-import and raise a warning/exception if this is not installed?

**Review by richardliaw — CHANGES_REQUESTED** [review]

File: python/setup.py (hunk 1 above)
Comment (brycehuang30) [reply python/setup.py:345]: sure, will do

**Comment by brycehuang30** [comment]
I updated the PR to soft import `debugpy`

Belowing are the manual tests after adding the soft import for debugpy
breakpoint without `debugpy` installed
image

breakpoint with `debugpy` installed
image


Post mortem without `debugpy` installed
image

post mortem with `debugpy` installed
image

**Comment by brycehuang30** [comment]
I tried throwing an exception when import error occurred, but it doesn't work well. The exception will be captured by Ray and drop the program into post-mortem mode, which will try to import debugpy again and throw another import error again... And end up with multiple levels of exception all saying there is an import error.

File: python/ray/util/debugpy.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+def _try_import(module):
+    try:
+        return importlib.import_module(module)
+    except ModuleNotFoundError:
+        log.error(
+            f"Module '{module}' cannot be loaded. "
+            f"Ray Debugger will not work without '{module}'. "
+            f"Install this module using 'pip install {module}' "
+        )
```
Comment (brycehuang30) [inline python/ray/util/debugpy.py:23]: I tried throwing an exception when import error occurred, but it doesn't work well. The exception will be captured by Ray and drop the program into post-mortem mode, which will try to import debugpy again and throw another import error again... And end up with multiple levels of exception all saying there is an import error. So I choose to just log the error message and skip debugger logic if the module is not available.

File: python/ray/util/debugpy.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def _try_import(module):
+    try:
+        return importlib.import_module(module)
+    except ModuleNotFoundError:
+        log.error(
+            f"Module '{module}' cannot be loaded. "
+            f"Ray Debugger will not work without '{module}'. "
+            f"Install this module using 'pip install {module}' "
```
Comment (richardliaw) [inline python/ray/util/debugpy.py:22]: make this specific, look for debugpy >= 1.8.0, and then raise warning if not found (and suggest to user to install debugpy==1.8.0)

File: python/ray/util/debugpy.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+POST_MORTEM_ERROR_UUID = "post_mortem_error_uuid"
+
+
+def _try_import(module):
+    try:
+        return importlib.import_module(module)
+    except ModuleNotFoundError:
+        log.error(
+            f"Module '{module}' cannot be loaded. "
+            f"Ray Debugger will not work without '{module}'. "
```
Comment (richardliaw) [inline python/ray/util/debugpy.py:21]: Ray Advanced Debugger

**Review by richardliaw — APPROVED** [review]

**Comment by MissiontoMars** [comment]
The RAY_DEBUG environment variable name is so confusing that users will try to use the ray debug command to debug it, but ray debug is actually used in conjunction with RAY_PDB. @brycehuang30 @richardliaw

**Pull request merged**